File: linux/QMKToolbox/Usb/Bootloader/Stm32DfuDevice.cs
// ReSharper disable StringLiteralTypo
using System.IO;
using System.Threading.Tasks;

namespace QMK_Toolbox.Usb.Bootloader;

internal class Stm32DfuDevice : BootloaderDevice
{
    public Stm32DfuDevice(KnownHidDevice d) : base(d)
    {
        Type = BootloaderType.Stm32Dfu;
        Name = "STM32 DFU";
        IsResettable = true;
    }

    public override void Flash(string mcu, string file)
    {
        if (Path.GetExtension(file)?.ToLower() == ".bin")
            RunProcessAsync("dfu-util", $"-a 0 -d 0483:DF11 -s 0x08000000:leave -D \"{file}\"").Wait();
        else
            PrintMessage("Only firmware files in .bin format can be flashed with dfu-util!", MessageType.Error);
    }

    public override void Reset(string mcu)
    {
        RunProcessAsync("dfu-util", "-a 0 -d 0483:DF11 -s 0x08000000:leave").Wait();
    }
}

File: linux/src/qmk_toolbox/bootloader/stm32_dfu_device.py
from .bootloader_device import BootloaderDevice
from .bootloader_type import BootloaderType


class Stm32DfuDevice(BootloaderDevice):
    def __init__(self, usb_device):
        super().__init__(usb_device)
        self.bootloader_type = BootloaderType.STM32_DFU
        self.name = "STM32 DFU"
        self.preferred_driver = "usbhid"
    
    async def flash(self, mcu: str, file_path: str) -> int:
        args = ["-d", f"{self.vendor_id:04x}:{self.product_id:04x}", "-a", "0", "-s", "0x08000000:leave", "-D", file_path]
        return await self._run_process("dfu-util", args)
